src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -266,9 +271,99 @@ withAssms :: InstEnv a -> ICtx -> Diff -> Maybe SubcId -> (InstEnv a -> ICtx ->
-    forM_ assms (SMT.smtAssert ieSMT)
-    act env' ctx' { icAssms = mempty }
+
+  let smtRewrites = if etabeta ieCfg then
+                      concatMap (\i -> extractRewrites $ lookupBindEnv i ieBEnv) delta
+                    else
+                      mempty
+
+  SMT.smtBracket ieSMT "PLE.evaluate" $ do
+    forM_ assms (SMT.smtAssert ieSMT )
+    act env' ctx' { icAssms = mempty, icSMTRewrites = M.union (M.fromList smtRewrites) (icSMTRewrites ctx) }
+  where
+
+    -- [(is$Something foo);
+    --  (~((is$SomethingElse foo));
+    --  ...
+    --  ((prop foo) = 
+    --    (Something v₁ ... vₙ));
+    --  ((prop foo) =
+    --    (Something m₁ ... mₙ));
+    --  ...]
+
+    -- We solve the system v₁ = m₁, v₂ = m₂, ... vₙ = mₙ
+    -- finding equalities between the free variables
+
+    -- Given ((Something v₁ ... vₙ), (Something m₁ ... mₙ))
+    -- or    (v₁,                    (Something m₁ ... mₙ))
+    -- obtain the unification
+    obtainEqualities (e1, e2)
+      | fst1 : args1 <- flatten $ dropECst e1
+      , fst2 : args2 <- flatten $ dropECst e2
+      , EVar cons1 <- dropECst fst1
+      , EVar cons2 <- dropECst fst2
+      , cons1 == cons2
+      , Just _ <- M.lookup cons1 (knDataCtors ieKnowl)
+      = unifyArgumets args1 args2
+    obtainEqualities (e1, e2)
+      | EVar f <- dropECst e1
+      , validVar f
+      , fst2 : _ <- flatten $ dropECst e2
+      , (EVar cons) <- dropECst fst2
+      , Just _ <- M.lookup cons $ knDataCtors ieKnowl
+      = Just [(f, e2)]
+    obtainEqualities (e1, e2)
+      | EVar f <- dropECst e2
+      , validVar f
+      , fst1 : _ <- flatten $ dropECst e1
+      , (EVar cons) <- dropECst fst1
+      , Just _ <- M.lookup cons $ knDataCtors ieKnowl
+      = Just [(f, e1)]
+    obtainEqualities _ = Nothing
+
+    flatten (EApp x xs) = flatten x ++ [ xs ]
+    flatten x = [x]
+
+    unifyArgumets [] [] = Just []
+    unifyArgumets (x:xs) (y:ys)
+      | EVar xn <- dropECst x
+      , validVar xn
+      = fmap ((xn, y) :) (unifyArgumets xs ys)
+    unifyArgumets (x:xs) (y:ys)
+      | EVar yn <- dropECst y
+      , validVar yn
+      = fmap ((yn, x) :) (unifyArgumets xs ys)
+    unifyArgumets (x:xs) (y:ys)
+      | dropECst x == dropECst y
+      = unifyArgumets xs ys
+    unifyArgumets _ _ 
+      = Nothing
+
+    -- Looks inside a list of qualities for equalities 
+    -- that share the lhs and then perform the unification
+    extractRewrites (_, sreft, _)
+      | Reft (_, PAnd ra) <- sr_reft sreft
+      =  concat
+      $  Mb.mapMaybe obtainEqualities
+      $  Mb.mapMaybe (keepTrans . snd)
+      $  Misc.groupList 
+      $  bimap unApply unApply
+     <$> Mb.mapMaybe keepPropEq ra
+    extractRewrites _ = []
+
+    keepPropEq (PAtom Eq e1 e2) = Just (e1, e2)
+    keepPropEq _ = Nothing
+    
+    keepTrans [a, b] = Just (a, b)
+    keepTrans _ = Nothing
+
+-- Find a way to make this check more robust
+-- Check if a symbol is a local variable
+validVar :: Symbol -> Bool
```

Comment:
My account of what this implements: users of LF are asked to mark variables that are considered for unification-induced rewriting by giving them special names (you called them local so far). The users are responsible for avoiding non-confluent rewriting rule sets that could result from these unifications.

It looks like a hard proposition. Firstly, predicting the rule sets that will be produced requires calculation. And second, determining termination is non-trivial.

A partial solution would be to do the unification in LH, and augment the fq file/info with rewrites that apply per constraint rather than globally. This solution is partial because it makes explicit the rewrites (good), but it still leaves to the user to ensure termination. On the other hand, maybe this isn't worse than asking the user to ensure termination of reflected functions.

Lastly, is requiring local variables on the left of a rewrite `(x, e)` sufficient to ensure termination? Maybe `e` should not contain local variables, or some other restriction of the sort.

Reply:
I think you can't have a constraint that isn't false that is cyclic since they are required to be the argument of a contractor, you're right about the naming thing but I couldn't find any other mechanisms, I would have expected variables to be tagged differently based on the context they are introduced like in ghc

Reply:
Another solution that maybe it's also retro compatible is having a list of all the definitions like we have with type constructors

Reply:
Then a local variable is anything that is not inside the definitions list && is not inside the ANF list

Reply:
That looks like a possible mechanism to mark variables. But let me stress that the user experience problem I'm surfacing is rather independent of how the local variables are signaled.

Reply:
Deferred by removal of the change.